cmd/tsbs_generate_queries/databases/clickhouse/devops_test.go

```diff
@@ -84,3 +88,443 @@ func TestDevopsGetSelectClausesAggMetrics(t *testing.T) {
+            FROM cpu
+            ORDER BY
+                hostname ASC,
+                created_at DESC
+            `,
+		},
+		{
+			desc:               "use tags",
+			devopsUseTags:      true,
+			expectedHumanLabel: "ClickHouse last row per host",
+			expectedHumanDesc:  "ClickHouse last row per host",
+			expectedQuery: `
+            SELECT *
+            FROM
+            (
+                SELECT *
+                FROM cpu
+                WHERE (tags_id, created_at) IN
+                (
+                    SELECT
+                        tags_id,
+                        max(created_at)
+                    FROM cpu
+                    GROUP BY tags_id
+                )
+            ) AS c
+            ANY INNER JOIN tags AS t ON c.tags_id = t.id
+            ORDER BY
+                t.hostname ASC,
+                c.time DESC
+            `,
+		},
+	}
+	testFunc := func(d *Devops, c testCase) query.Query {
+		q := d.GenerateEmptyQuery()
+		d.LastPointPerHost(q)
+		return q
+	}
+
+	start := time.Unix(0, 0)
+	end := start.Add(devops.HighCPUDuration).Add(time.Hour)
+
+	runTestCases(t, testFunc, start, end, cases)
+}
+
+func TestGroupByTime(t *testing.T) {
+	cases := []testCase{
+		{
+			desc:               "happy path",
+			input:              1,
+			expectedHumanLabel: "ClickHouse 1 cpu metric(s), random    1 hosts, random 1s by 1m",
+			expectedHumanDesc:  "ClickHouse 1 cpu metric(s), random    1 hosts, random 1s by 1m: 1970-01-01T01:09:26Z",
+			expectedQuery: `
+        SELECT
+            toStartOfMinute(created_at) AS minute,
+            max(usage_user) AS max_usage_user
+        FROM cpu
+        WHERE (hostname = 'host_9') AND (created_at >= '1970-01-01 01:09:26') AND (created_at < '1970-01-01 01:09:27')
+        GROUP BY minute
+        ORDER BY minute ASC
+        `,
+		},
+	}
+
+	testFunc := func(d *Devops, c testCase) query.Query {
+		q := d.GenerateEmptyQuery()
+		d.GroupByTime(q, c.input, 1, time.Second)
+		return q
+	}
+
+	start := time.Unix(0, 0)
+	end := start.Add(2 * time.Hour)
+
+	runTestCases(t, testFunc, start, end, cases)
+}
+
+type testCase struct {
+	desc               string
+	input              int
+	devopsUseTags      bool
+	fail               bool
+	failMsg            string
+	expectedHumanLabel string
+	expectedHumanDesc  string
+	expectedQuery      string
+}
+
+func runTestCases(t *testing.T, testFunc func(*Devops, testCase) query.Query, s time.Time, e time.Time, cases []testCase) {
+	rand.Seed(123) // Setting seed for testing purposes.
+
+	for _, c := range cases {
+		t.Run(c.desc, func(t *testing.T) {
+			d := NewDevops(s, e, 10)
+			d.UseTags = c.devopsUseTags
+
+			if c.fail {
+				func() {
+					defer func() {
+						r := recover()
+						if r == nil {
+							t.Errorf("did not panic when should")
+						}
+
+						if r != c.failMsg {
+							t.Fatalf("incorrect fail message: got %s, want %s", r, c.failMsg)
+						}
+					}()
+
+					testFunc(d, c)
+				}()
+				return
```

Comment:
I think it's probably better to just wrap the other case in an `else` rather than early return. it's more obvious that they are two different paths that way